src/coordinator/ts/m3db/convert.go

```diff
@@ -0,0 +1,138 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"io"
+	"time"
+
+	"github.com/m3db/m3db/src/dbnode/encoding"
+	"github.com/m3db/m3db/src/dbnode/encoding/m3tsz"
+	"github.com/m3db/m3db/src/dbnode/x/xio"
+)
+
+// BlockReplica contains the replicas for a single m3db block
+type BlockReplica struct {
+	Start     time.Time
+	BlockSize time.Duration
+	Replicas  []encoding.MultiReaderIterator
+}
+
+func convertM3DBSeriesIterators(iterators encoding.SeriesIterators) ([]SeriesBlocks, error) {
+	iterAlloc := func(r io.Reader) encoding.ReaderIterator {
+		iter := m3tsz.NewDecoder(true, encoding.NewOptions())
+		return iter.Decode(r)
+	}
+
+	var multiSeriesBlocks []SeriesBlocks
+
+	for _, seriesIterator := range iterators.Iters() {
+
+		blockReplicas, err := blockReplicasFromSeriesIterator(seriesIterator, iterAlloc)
+		if err != nil {
+			return []SeriesBlocks{}, err
+		}
+
+		series := seriesBlocksFromBlockReplicas(blockReplicas, seriesIterator)
+
+		multiSeriesBlocks = append(multiSeriesBlocks, series)
+	}
+
+	return multiSeriesBlocks, nil
+}
+
+func blockReplicasFromSeriesIterator(seriesIterator encoding.SeriesIterator, iterAlloc encoding.ReaderIteratorAllocate) ([]BlockReplica, error) {
+	var blockReplicas []BlockReplica
+	for _, replica := range seriesIterator.Replicas() {
+		perBlockSliceReaders := replica.Readers()
+		next := true
+		for next {
+			l, start, bs := perBlockSliceReaders.CurrentReaders()
+			var readers []xio.SegmentReader
+			for i := 0; i < l; i++ {
+				reader := perBlockSliceReaders.CurrentReaderAt(i)
+				// import to clone the reader as we need its position reset before
+				// we use the contents of it again
+				clonedReader, err := reader.Clone()
+				if err != nil {
+					return []BlockReplica{}, err
+				}
+				readers = append(readers, clonedReader)
+			}
+			// todo(braskin): pooling
+			iter := encoding.NewMultiReaderIterator(iterAlloc, nil)
+			iter.Reset(readers, start, bs)
+
+			inserted := false
+			for i := range blockReplicas {
```

Comment:
Would using a map with startTime as the key like `[time.Time][]BlockReplica` be better?

Alternatively, can do some maths using the Start and BlockSize to figure out which index this should go in?

Reply:
Per our convo, I'll hold off on this for now.